tests/functional/history/test_db.py

```diff
@@ -0,0 +1,199 @@
+            # Normally our reader would take care of parsing the JSON from
+            # the payload.
+            if col_name == 'payload':
+                record[col_name] = json.loads(record[col_name])
+            self.assertEqual(record[col_name], row_value)
+
+        self.assertIn('id', record.keys())
+        self.assertIn('request_id', record.keys())
+
+    def test_can_write_many_records(self):
+        writer = DatabaseRecordWriter(connection=self.connection)
+        known_record_fields = {
+            'source': 'TEST',
+            'event_type': 'foo',
+            'payload': '',
+            'timestamp': 12334
+        }
+        records_to_write = 40
+        for _ in range(records_to_write):
+            writer.write_record(known_record_fields)
+
+        cursor = self._get_cursor()
+        cursor.execute("SELECT COUNT(*) FROM records;")
+        num_records = cursor.fetchone()
+        self.assertEqual(num_records[0], records_to_write)
+
+
+@unittest.skipIf(sqlite3 is None,
+                 "sqlite3 not supported in this python")
+class TestDatabaseRecordReader(BaseDatabaseTest):
+    def _write_sequence_of_records(self, writer, records):
+        for record in records:
+            writer.write_record(record)
+
+    def test_yields_nothing_if_no_matching_record_id(self):
+        reader = DatabaseRecordReader(self.connection)
+        records = [record for record in reader.yield_records('fake_id')]
+        self.assertEqual(len(records), 0)
+
+    def test_yields_nothing_no_recent_records(self):
+        reader = DatabaseRecordReader(self.connection)
+        records = [record for record in reader.yield_latest_records()]
+        self.assertEqual(len(records), 0)
+
+    def test_can_read_record(self):
+        writer_a = DatabaseRecordWriter(self.connection)
+        writer_b = DatabaseRecordWriter(self.connection)
+        self._write_sequence_of_records(writer_a, [
+            {
+                'source': 'TEST',
+                'event_type': 'foo',
+                'payload': '',
+                'timestamp': 3
+            },
+            {
+                'source': 'TEST',
+                'event_type': 'bar',
+                'payload': '',
+                'timestamp': 1
+            },
+            {
+                'source': 'TEST',
+                'event_type': 'baz',
+                'payload': '',
+                'timestamp': 4
+            }
+        ])
+        self._write_sequence_of_records(writer_b, [
+            {
+                'source': 'TEST',
+                'event_type': 'qux',
+                'payload': '',
+                'timestamp': 2
+            },
+            {
+                'source': 'TEST',
+                'event_type': 'zip',
+                'payload': '',
+                'timestamp': 6
+            }
+        ])
+        reader = DatabaseRecordReader(self.connection)
+        cursor = self._get_cursor()
+        cursor.execute(
```

Comment:
I wonder if it makes sense to expose the ability to supply your own id to the writer intializer instead of having to do this? Then you can test for the other id pretty easily as well.

Reply:
Yea this was a little annoying but I didn't see any reason to ever need that outside tests so I left it as is. In the event that is ever used in practice it could easy lead to colliding ids in a database which means you would start picking up records that didn't belong together.

Reply:
Yeah that makes sense. It was just a thought I had when I was looking through the tests.